test/e2e/aws_util_test.go

```diff
@@ -0,0 +1,312 @@
+		if err != nil {
+			return nil, fmt.Errorf("Error describing Route Tables: %v", err)
+		}
+
+		for _, rt := range result.RouteTables {
+			for _, assoc := range rt.Associations {
+				if assoc != nil && assoc.SubnetId != nil {
+					publicSubnets[aws.StringValue(assoc.SubnetId)] = struct{}{}
+				}
+			}
+		}
+	}
+	return publicSubnets, nil
+}
+
+// createSession creates a session using decoded AWS credentials.
+func createSession(region, decodedAccessKeyID, decodedSecretAccessKey string) (*session.Session, error) {
+	sess, err := session.NewSession(&aws.Config{Region: aws.String(region),
+		Credentials: credentials.NewStaticCredentials(string(decodedAccessKeyID), string(decodedSecretAccessKey), "")})
+	if err != nil {
+		return nil, fmt.Errorf("Error creating session: %v", err)
+	}
+	return sess, nil
+}
+
+// getVPCID fetched AWS VPC ID tagged with cluster name as tag key.
+func getVPCID(clusterName string, svc *ec2.EC2) (string, error) {
+	// Define the filter to search for instances with the specified tag
+	tagKey := "kubernetes.io/cluster/" + clusterName
+	tagValue := "owned"
+	filters := []*ec2.Filter{
+		{
+			Name:   aws.String("tag:" + tagKey),
+			Values: []*string{aws.String(tagValue)},
+		},
+	}
+
+	// Describe instances that match the filter
+	result, err := svc.DescribeInstances(&ec2.DescribeInstancesInput{
+		Filters: filters,
+	})
+	if err != nil {
+		return "", fmt.Errorf("Error describing instances: %v", err)
+	}
+
+	// Check if any instances were found
+	if len(result.Reservations) == 0 || len(result.Reservations[0].Instances) == 0 {
+		return "", fmt.Errorf("No instances found with the specified tag, or unable to determine VPC ID.")
+	}
+	instances := result.Reservations[0].Instances
+	for _, instance := range instances {
+		if instance.VpcId != nil {
+			// Extract and return the VPC ID
+			return aws.StringValue(instance.VpcId), nil
+		}
+	}
+	return "", fmt.Errorf("Can't find VPC ID due to error: %v", err)
+}
+
+// cleanupEIPAllocations clean all unassociated EIPs created during the e2e tests.
+func cleanupEIPAllocations(t *testing.T, svc *ec2.EC2, infraConfig configv1.Infrastructure) bool {
+	t.Log("Release unassociated EIPs ")
+
+	clusterName, err := getClusterName(infraConfig)
+	if err != nil {
+		t.Fatalf("failed to get cluster name due to error: %v", err)
+	}
+
+	vpcID, err := getVPCID(clusterName, svc)
+	if err != nil {
+		t.Fatalf("failed to get VPC ID due to error %s", err)
+	}
+
+	tagKeyEIP := fmt.Sprintf("NE1674-%s", t.Name())
+	tagValueEIP := clusterName
+
+	// Describe addresses to get unassociated EIPs
+	describeAddressesInput := &ec2.DescribeAddressesInput{}
+	describeAddressesOutput, err := svc.DescribeAddresses(describeAddressesInput)
+	if err != nil {
+		t.Fatalf("failed to describe addresses, %v", err)
+	}
+
+	var unassociatedEIPs []struct {
+		PublicIp     string
+		AllocationId string
+	}
+
+	var associatedEIPs []struct {
+		PublicIp     string
+		AllocationId string
+	}
```

Comment:
Why not already use the struct exists for addresses? It'll make your code more clean & readable.

```suggestion
	var unassociatedEIPs, associatedEIPs []*ec2.Address
```